File: cli/src/lib/seed.js
export class Seed {
  constructor() {
    this.domain = '';
    this.obj_vocab = [];
    this.attr_vocab = [];
    this.obj_tree = {};
    this.attr_tree = {};
    this.incidence_hints = {};
    this.conventions = [];
    this.reference_k = null;
  }

  lookup_obj(name) {
    if (name in this.obj_tree) return this.obj_tree[name];
    for (const [k, v] of Object.entries(this.obj_tree)) {
      if (k.includes(name) || name.includes(k)) return v;
    }
    return null;
  }

  lookup_attr(name) {
    if (name in this.attr_tree) return this.attr_tree[name];
    for (const [k, v] of Object.entries(this.attr_tree)) {
      if (k.includes(name) || name.includes(k)) return v;
    }
    return null;
  }

  has_content() {
    return !!(Object.keys(this.obj_tree).length || Object.keys(this.attr_tree).length ||
              this.obj_vocab.length || this.attr_vocab.length ||
              this.conventions.length || this.reference_k);
  }

  conventions_text() {
    return this.conventions.map(c => `- ${c}`).join('\n');
  }

  to_dict() {
    return {
      domain: this.domain,
      obj_vocab: this.obj_vocab,
      attr_vocab: this.attr_vocab,
      obj_tree: this.obj_tree,
      attr_tree: this.attr_tree,
      incidence_hints: this.incidence_hints,
      conventions: this.conventions,
      reference_k: this.reference_k
    };
  }

  from_dict(d) {
    this.domain = d.domain || '';
    this.obj_vocab = d.obj_vocab || [];
    this.attr_vocab = d.attr_vocab || [];
    this.obj_tree = d.obj_tree || {};
    this.attr_tree = d.attr_tree || {};
    this.incidence_hints = d.incidence_hints || {};
    this.conventions = d.conventions || [];
    this.reference_k = d.reference_k || null;
  }
}

File: konceptos/seed.py
"""KonceptOS v2.1 — Seed: pluggable prior knowledge at decision points."""
from .util import VALID_I

class Seed:
    """Base interface. Return None = 'I don't know'."""
    def suggest_direction(self,on,od,an,ad,ctx=None): return None
    def suggest_split(self,name,desc,kind,ctx=None): return None
    def suggest_schema(self,an,ad): return None

class JsonSeed(Seed):
    """Seed backed by JSON with vocab, trees, hints, conventions."""
    def __init__(self):
        self.domain="";self.obj_vocab=[];self.attr_vocab=[]
        self.obj_tree={};self.attr_tree={}
        self.incidence_hints={};self.conventions=[];self.reference_k=None

    def suggest_direction(self,on,od,an,ad,ctx=None):
        h=self.incidence_hints.get('%s|%s'%(on,an))
        if not h: h=self.incidence_hints.get('*|%s'%an)
        if not h: h=self.incidence_hints.get('%s|*'%on)
        return h.upper() if h and h.upper() in VALID_I else None

    def suggest_split(self,name,desc,kind,ctx=None):
        tree=self.obj_tree if kind=='obj' else self.attr_tree
        ch=tree.get(name)
        if not ch:
            for k,v in tree.items():
                if k in name or name in k: ch=v;break
        return [{'name':c,'desc':''} for c in ch] if ch else None

    def has_content(self):
        return bool(self.obj_tree or self.attr_tree or self.obj_vocab or self.conventions)

    def to_dict(self):
        return {'domain':self.domain,'obj_vocab':self.obj_vocab,'attr_vocab':self.attr_vocab,
                'obj_tree':self.obj_tree,'attr_tree':self.attr_tree,
                'incidence_hints':self.incidence_hints,'conventions':self.conventions,
                'reference_k':self.reference_k}

    def from_dict(self,d):
        self.domain=d.get('domain','');self.obj_vocab=d.get('obj_vocab',[])
        self.attr_vocab=d.get('attr_vocab',[]);self.obj_tree=d.get('obj_tree',{})
        self.attr_tree=d.get('attr_tree',{});self.incidence_hints=d.get('incidence_hints',{})
        self.conventions=d.get('conventions',[]);self.reference_k=d.get('reference_k')

    def summary(self):
        p=['  Seed: %s'%(self.domain or '(unnamed)')]
        if self.obj_vocab: p.append('    obj vocab: %d'%len(self.obj_vocab))
        if self.attr_vocab: p.append('    attr vocab: %d'%len(self.attr_vocab))
        if self.obj_tree: p.append('    obj tree: %d'%len(self.obj_tree))
        if self.attr_tree: p.append('    attr tree: %d'%len(self.attr_tree))
        if self.incidence_hints: p.append('    hints: %d'%len(self.incidence_hints))
        if self.conventions: p.append('    conventions: %d'%len(self.conventions))
        return '\n'.join(p)

class SeedChain(Seed):
    """Priority chain. First non-None answer wins."""
    def __init__(self,seeds=None): self.seeds=seeds or []
    def add(self,s): self.seeds.append(s)
    def suggest_direction(self,on,od,an,ad,ctx=None):
        for s in self.seeds:
            r=s.suggest_direction(on,od,an,ad,ctx)
            if r is not None: return r
        return None
    def suggest_split(self,name,desc,kind,ctx=None):
        for s in self.seeds:
            r=s.suggest_split(name,desc,kind,ctx)
            if r is not None: return r
        return None
